### DISCUSSION TITLE
Early CLOG Builder screen shot

### DISCUSSION THREAD

**Discussion by rabbibotton** [rabbibotton/clog#89] [Show and tell]
![Screenshot (2)](https://user-images.githubusercontent.com/11616789/144544233-47edc4f9-e7f6-4aaa-a3a6-ae4cca60fc75.png)
 
My goal is Jan 1st something ready for real use.

David Botton

**Comment by aykaramba** [comment]
Love it!

**Comment by kaeland** [comment]
This looks cool!

**Reply by rabbibotton** [reply]
It is coming along amazing :) while not ready to build anything yet

git checkout builder
git pull
(ql:quickload :clog/tools) (clog-tools:clog-builder)

**Comment by rabbibotton** [comment]
The builder now is actually functioning, granted only has text boxes, buttons and labels at the moment, and can only render a GUI to code, but it is actually functional :)

**Reply by kaeland** [reply]
Awesome!

**Reply by kaeland** [reply]
Dr. Botton, just wondering, but how would you like to see this project grow? Are there tools or features that a community of others could help you build along the way?

**Reply by rabbibotton** [reply]
For sure! The real boost for CLOG and the CLOG builder will come with creating more functional widgets. There is documentation on it in the tutorials but I know more needed. The CLOG builder will be extensible to easily add them from the start. Once I have a bit more of the Builder done I'll try and demonstrate that more.

Some plans for the builder
1) In addition to the fixed locating of widgets various ways to do layouts based on flows (flexbox type layouts)
2) No code GUIs and Websites, ie you click on say a button and then choose actions like switch another frame, save fields to a database, etc
3) Something to generate reports
4) Batch processing wizards

Another area is sample apps, best way for you and others to grow with CLOG. Perhaps a tetris game, websites like a shopping cart, etc.

**Reply by aykaramba** [reply]
As well, technically any appropriate package in quicklisp could receive a CLOG frontend as appropriate.  

What would really be interesting is if someone wrote a CLOG interface for LEM (https://github.com/lem-project/lem).

**Reply by rabbibotton** [reply]
I actually looked at that today (CLOG + LEM) and is a good fit to integrate with the builder. So I may consider doing it at some point (or maybe someone else will :)

**Comment by rabbibotton** [comment]
![Screenshot (3)](https://user-images.githubusercontent.com/11616789/145912101-6543cf19-0a55-4772-9d4d-15f0ef0c134b.png)

**Reply by rabbibotton** [reply]
It is trivial to add now more widgets, just not sure have the strength tonight. The generated code works but needs work in the sense that each form I think I will go with it being its own package, everything encapsulated, no globals etc.

**Reply by aykaramba** [reply]
Okay that is awesome.  

Not quite related, but, your approach here will make it easy to create a simple wordpress style cms and then allow us to rebuild the wordprss layout widgets from Elementor (https://github.com/elementor/elementor).

Quick demo of how it works in Wordpress is on their home page:  https://elementor.com/

I will be trying to carve out solid hours to upgrade my skills so I can start trying to build a simplified version of that.   Actually, what I really want to do is steal some of the ideas from Wordpress + Elementor and then build something like the Hugo static site builder, as per: https://gohugo.io/

Now that we have figured how to dump CLOG to an image, having a nice local binary replicating a basic Hugo style static site generator should be awesome.  If someone were to create a CLOG frontend for LEM, static site generation could get REALLY interesting really quickly especially if we start to think about LEM extensions that are CLOG aware.

This is a really exciting.

**Comment by rabbibotton** [comment]
I for sure plan on being able to layout a whole gui/website in builder and just dump an executable at some point ready to deploy. These other products are focused on delivering websites, my idea is to deliver full applications in a click not just a website (but of course that too).

**Reply by aykaramba** [reply]
Since you mentioned it ... 

Dr. Botton, I just wanted to draw your attention to this Reddit thread briefly: https://old.reddit.com/r/lisp/comments/r2s2iu/android_lispworks_weblocks_demo_application/

Somone used the Lispworks implementation ability to compile Weblocks to an Android mobile app and run a local (to the Android device) weblocks server + web interface.  Obviously, CLOG technically now allows us to provide web frontends for lots of software using the Lispworks tooling.

Neat, eh?

Awesome stuff.

**Reply by rabbibotton** [reply]
My Ada vs of CLOG has been used many times in iPhone and Android apps. It works well either local or remote. You currently can use ecl to deploy on iPhone and Android.

To deploy a server based CLOG app you can use https://cordova.apache.org/ on the boot.html and boot.js file.
To deploy as a phone/tablet app you create your app using the native toolkit (xcode or android's) and link in ECL.

I write CLOG in 5-10 minute bites of time between patients, teaching etc wish had the time to show it all off.

**Reply by aykaramba** [reply]
That is amazing.  Thank you kindly for your time and this contribution.  Very inspiring.

**Reply by rabbibotton** [reply]
BTW CLOG runs fantastic in ECL (including the builder). I regularly test on SBCL and ECL. They are amazing tools. I do not own a copy of LW so haven't used much but I have made changes in CLOG to allow for its use on LW and one or two have expressed some interest in working to make sure works well on it also.

**Reply by aykaramba** [reply]
I am obviously wasting your time with the LW link, my apologies.  I now have yet more to learn.   Time to install ECL as well as look at Cordova.  

Amazing.

**Reply by rabbibotton** [reply]
Not a waste of my time at all, I didn't know and happy to hear about it.
